Integrations/src/main/java/io/deephaven/integrations/python/PythonReplayListenerAdapter.java

```diff
@@ -69,18 +77,36 @@ public void replay() {
         final TableUpdate update =
                 new TableUpdateImpl(source.getRowSet(), emptyRowSet, emptyRowSet, emptyShift, emptyColumnSet);
         final boolean isReplay = true;
-        pyCallable.call("__call__", update, isReplay);
+        pyListenerCallable.call("__call__", update, isReplay);
     }
 
     @Override
     public void onUpdate(final TableUpdate update) {
         final boolean isReplay = false;
-        pyCallable.call("__call__", update, isReplay);
+        pyListenerCallable.call("__call__", update, isReplay);
+    }
+
+    @Override
+    public void onFailureInternal(Throwable originalException, Entry sourceEntry) {
+        if (!pyOnFailureCallback.isNone()) {
+            try {
+                pyOnFailureCallback.call("__call__", ExceptionUtils.getStackTrace(originalException));
+            } catch (Throwable e) {
+                // If the Python onFailure callback fails, log the new exception
+                // and continue with the original exception.
+                log.error().append("Python on_error callback failed: ").append(e).endl();
+            }
+        }
+        super.onFailureInternal(originalException, sourceEntry);
     }
 
     @Override
     public boolean canExecute(final long step) {
         return super.canExecute(step)
                 && (dependencies.length == 0 || Arrays.stream(dependencies).allMatch(t -> t.satisfied(step)));
     }
+
+    public boolean isFailed() {
```

Comment:
It is OK with me if you want to introduce a `failed` boolean in `MergedListener`, and an `isFailed()` method there. Set `failed = true` in `propagateError()`. You could also check it and refuse to schedule notifications in the future, but that shouldn't *really* be necessary.

Code snippet:
```
    public final void notifyChanges() {
        notifyInternal(null, null);
    }

    private void notifyInternal(@Nullable final Throwable upstreamError,
            @Nullable final TableListener.Entry errorSourceEntry) {
        if (failed) {
            return;
        }
```